contracts/verifiers/RequestOwnership.sol

```diff
@@ -24,48 +24,40 @@ abstract contract RequestOwnership is ZKPVerifierBase {
         }
     }
 
-    /// @dev Modifier to check if the caller is ZKP Request owner
-    modifier onlyRequestOwner(uint64 requestId) virtual {
-        require(getRequestOwner(requestId) == _msgSender(), "Not a request owner");
-        _;
-    }
-
-    /// @dev Sets a ZKP request
-    /// @param requestId The ID of the ZKP request
-    /// @param request The ZKP request data
-    function setZKPRequest(
-        uint64 requestId,
-        IZKPVerifier.ZKPRequest calldata request
-    ) public virtual override {
-        super.setZKPRequest(requestId, request);
-        _setRequestOwner(requestId, _msgSender());
-    }
-
     /**
-     * @dev Set the list of ZKP requests for the list of requestIds in the same order.
-     * @param requestIds Request ids of the ZKP requests.
-     * @param requests ZKP requests to set.
+     * @dev Sets different requests
+     * @param singleRequests The requests that are not in any group
+     * @param groupedRequests The requests that are in a group
      */
-    function setZKPRequests(
-        uint64[] calldata requestIds,
-        ZKPRequest[] calldata requests
+    function setRequests(
+        IVerifier.Request[] calldata singleRequests,
+        IVerifier.GroupedRequests[] calldata groupedRequests
     ) public virtual override {
-        for (uint256 i = 0; i < requestIds.length; i++) {
-            setZKPRequest(requestIds[i], requests[i]);
+        super.setRequests(singleRequests, groupedRequests);
+        for (uint256 i = 0; i < singleRequests.length; i++) {
+            _setRequestOwner(singleRequests[i].requestId, _msgSender());
+        }
+
+        for (uint256 i = 0; i < groupedRequests.length; i++) {
+            for (uint256 j = 0; j < groupedRequests[i].requests.length; j++) {
+                _setRequestOwner(groupedRequests[i].requests[j].requestId, _msgSender());
+            }
         }
     }
```

Comment:
Maybe overriding  _setRequest and adding _setRequestOwner(singleRequests[i].requestId, _msgSender()); to its logic may help avoid these two cycles